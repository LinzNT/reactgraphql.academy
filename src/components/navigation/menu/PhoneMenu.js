import { slide as Menu } from 'react-burger-menu'
import styled from 'styled-components'
import Link from 'gatsby-link'
import React from 'react'
import MenuData from './Menu.json'
import './PhoneMenu.css'

const Item = ({ children, ...props }) => <Link {...props}>{children}</Link>

const PhoneMenuItem = styled(Item)`
  display: block;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
`

PhoneMenuItem.displayName = 'PhoneMenuItem'

const PhoneMenu = () => (
  <Menu>
    {MenuData.map((item, i) => (
      <PhoneMenuItem key={i} to={item.to}>
        {item.text}
      </PhoneMenuItem>
    ))}
  </Menu>
)

export default PhoneMenu
