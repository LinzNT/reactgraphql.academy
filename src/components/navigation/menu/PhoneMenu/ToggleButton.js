import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { WHITE } from '../../../../config/styles'

const ToggleButton = ({ toggleMenu, className, isOpen }) => (
  <div className={className}>
    <input
      type="checkbox"
      name="main-menu-button"
      id="main-menu-button"
      checked={isOpen}
      onClick={toggleMenu}
      role="button"
      aria-pressed={isOpen.toString()}
      aria-labelledby="main-menu-button-label"
    />
    <span />
    <span />
    <span />
    <label id="main-menu-button-label" for="main-menu-button">
      Menu
    </label>
  </div>
)

const StyledToggleButton = styled(ToggleButton)`
  position: fixed;
  width: 36px;
  height: 30px;
  right: 30px;
  top: 30px;
  z-index: 9991;

  label {
    font-size: 12px;
    color: ${WHITE};
  }

  input {
    display: block;
    width: 40px;
    height: 60px;
    position: absolute;
    top: -7px;
    left: -5px;

    cursor: pointer;

    opacity: 0; /* hide this */
    z-index: 2; /* and place it over the hamburger */

    -webkit-touch-callout: none;
  }

  span {
    display: block;
    width: 33px;
    height: 4px;
    margin-bottom: 5px;
    position: relative;

    background: ${WHITE};
    border-radius: 3px;

    z-index: 1;

    transform-origin: 4px 0px;

    transition: transform 0.5s cubic-bezier(0.77, 0.2, 0.05, 1),
      background 0.5s cubic-bezier(0.77, 0.2, 0.05, 1), opacity 0.55s ease;
  }

  span:first-child {
    transform-origin: 0% 0%;
  }
  span:nth-last-child(2) {
    transform-origin: 0% 100%;
  }
  input:checked ~ span {
    opacity: 1;
    transform: rotate(45deg) translate(5px, -5px);
  }
  input:checked ~ span:nth-last-child(3) {
    opacity: 0;
    transform: rotate(0deg) scale(0.2, 0.2);
  }
  input:checked ~ span:nth-last-child(2) {
    transform: rotate(-45deg) translate(3px, 5px);
  }
`

StyledToggleButton.propTypes = {
  toggleMenu: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  className: PropTypes.string,
}

export default StyledToggleButton
