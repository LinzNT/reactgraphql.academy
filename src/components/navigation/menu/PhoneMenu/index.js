import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Link from '../../Link'
import MenuData from '../Menu.json'
import { RJSALogo } from '../../../logos/RJSALogo'
import ToggleButton from './ToggleButton'

const Item = ({ children, ...props }) => <Link {...props}>{children}</Link>

export const PhoneMenuItem = styled(Item)`
  display: block;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
`
PhoneMenuItem.displayName = 'PhoneMenuItem'

export const Overlay = styled.div`
  top: 0;
  left: 0;
  z-index: 99999;
  position: fixed;
  z-index: 1000;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.3);
  opacity: 1;
  transition: opacity 0.3s ease 0s;
`
Overlay.displayName = 'Overlay'

export const MenuContent = styled.div`
  top: 0;
  left: 0;
  position: fixed;
  right: inherit;
  z-index: 1100;
  width: ${({ width }) => width}px;
  margin-left: ${({ isOpen, width }) => (isOpen ? 0 : `-${width}px`)};
  height: 100%;
  transition: all 0.5s ease 0s;
  background: #373a47;
  padding: 2.5em 1.5em 0;
  box-sizing: border-box;
  overflow: auto;
  color: #b8b7ad;
`
MenuContent.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  width: PropTypes.number.isRequired,
}

const PhoneMenu = ({ width, defaultIsOpen = false }) => {
  const [isOpen, setIsOpen] = useState(defaultIsOpen)
  const items = MenuData.reduce((accumulatedItems, currentItem) => {
    if (currentItem.children) {
      return [
        ...accumulatedItems,
        ...currentItem.children.filter(i => !i.hideOnMobile),
      ]
    } else if (!currentItem.hideOnMobile) {
      return [...accumulatedItems, currentItem]
    } else {
      return accumulatedItems
    }
  }, [])

  function toggleMenu() {
    setIsOpen(!isOpen)
  }

  return (
    <React.Fragment>
      <ToggleButton toggleMenu={toggleMenu} isOpen={isOpen} />
      {isOpen && <Overlay onClick={toggleMenu} />}
      <MenuContent width={width} isOpen={isOpen}>
        <RJSALogo />
        {items.map(item => (
          <PhoneMenuItem onClick={toggleMenu} key={item.to} to={item.to}>
            {item.text}
          </PhoneMenuItem>
        ))}
      </MenuContent>
    </React.Fragment>
  )
}

PhoneMenu.propTypes = {
  defaultIsOpen: PropTypes.bool,
  width: PropTypes.number,
}

PhoneMenu.defaultProps = {
  width: 300,
}

export default PhoneMenu
