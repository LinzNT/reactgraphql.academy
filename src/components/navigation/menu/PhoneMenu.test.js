import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import PhoneMenu, { Menu, PhoneMenuItem } from './PhoneMenu'

describe('<PhoneMenu />', () => {
  it('should close the menu when a menu item is clicked', () => {
    const wrapper = shallow(<PhoneMenu isOpen />)

    expect(wrapper.find(Menu).props().isOpen).toBe(true)

    wrapper
      .find(PhoneMenuItem)
      .first()
      .simulate('click')

    expect(wrapper.find(Menu).props().isOpen).toBe(false)
  })

  it('should be closed by default', () => {
    const wrapper = shallow(<PhoneMenu />)
    expect(wrapper.find(Menu).props().isOpen).toBe(false)
  })

  it('should render the list of menu items', () => {
    const wrapper = shallow(<PhoneMenu />)

    expect(wrapper.find(PhoneMenuItem).length).toBe(11)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render curriculum as the first item on the menu', () => {
    const wrapper = shallow(<PhoneMenu />)

    expect(
      wrapper
        .find(PhoneMenuItem)
        .first()
        .props()
        .children.toLowerCase()
    ).toMatch(/curriculum/)
  })
})
