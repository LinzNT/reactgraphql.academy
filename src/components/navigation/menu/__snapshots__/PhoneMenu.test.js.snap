// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneMenu /> should render the list of menu items 1`] = `
<Menu
  bodyClassName=""
  burgerBarClassName=""
  burgerButtonClassName=""
  className=""
  crossButtonClassName=""
  crossClassName=""
  disableCloseOnEsc={false}
  id=""
  isOpen={false}
  itemClassName=""
  itemListClassName=""
  menuClassName=""
  morphShapeClassName=""
  noOverlay={false}
  onStateChange={[Function]}
  outerContainerId=""
  overlayClassName=""
  pageWrapId=""
  styles={Object {}}
  width={300}
>
  <RJSALogo />
  <PhoneMenuItem
    key="0"
    onClick={[Function]}
    to="/curriculum"
  >
    Curriculum
  </PhoneMenuItem>
  <PhoneMenuItem
    key="1"
    onClick={[Function]}
    to="/react-redux-graphql-bootcamp"
  >
    React bootcamp
  </PhoneMenuItem>
  <PhoneMenuItem
    key="2"
    onClick={[Function]}
    to="/advanced-react-redux-graphql-bootcamp"
  >
    Advanced React
  </PhoneMenuItem>
  <PhoneMenuItem
    key="3"
    onClick={[Function]}
    to="/react-native-bootcamp"
  >
    React Native
  </PhoneMenuItem>
  <PhoneMenuItem
    key="4"
    onClick={[Function]}
    to="/react-redux-graphql-part-time-course"
  >
    Part-time
  </PhoneMenuItem>
  <PhoneMenuItem
    key="5"
    onClick={[Function]}
    to="/about-us"
  >
    About us
  </PhoneMenuItem>
  <PhoneMenuItem
    key="6"
    onClick={[Function]}
    to="mailto:hello@reactjs.academy"
  >
    Contact
  </PhoneMenuItem>
</Menu>
`;
